R/sampling.R

```diff
@@ -146,3 +144,233 @@ sample_combinations <- function(ntrain, ntest, nsamples, joint_sampling = TRUE)
+#'
+#' @param x_test Matrix, data.frame or data.table with the features of the observation whose
+#' predictions ought to be explained (test data). Dimension \code{1xp} or \code{px1}.
+#'
+#' @param x_train Matrix, data.frame or data.table with training data.
+#'
+#' @param p Positive integer. The number of features.
+#'
+#' @param sample Boolean. True indicates that the method samples from the terminal node
+#' of the tree whereas False indicates that the method takes all the samples if it is less than n_samples.
+#'
+#' @return data.table with \code{n_samples} (conditional) Gaussian samples
+#'
+#' @keywords internal
+#'
+#' @examples
+#' m <- 10
+#' n <- 40
+#' n_samples <- 50
+#' mu <- rep(1, m)
+#' cov_mat <- cov(matrix(rnorm(n * m), n, m))
+#' x_train <- data.table::data.table(MASS::mvrnorm(n, mu, cov_mat))
+#' x_test <- MASS::mvrnorm(1, mu, cov_mat)
+#' x_test_dt <- data.table::setDT(as.list(x_test))
+#' given_ind <- c(4, 7)
+#' dependent_ind <- (1:dim(x_train)[2])[-given_ind]
+#' x <- x_train[, given_ind, with = FALSE]
+#' y <- x_train[, dependent_ind, with = FALSE]
+#' df <- data.table::data.table(cbind(y, x))
+#' colnames(df) <- c(paste0("Y", 1:ncol(y)), paste0("V", given_ind))
+#' ynam <- paste0("Y", 1:ncol(y))
+#' fmla <- as.formula(paste(paste(ynam, collapse = "+"), "~ ."))
+#' datact <- party::ctree(fmla, data = df, controls = party::ctree_control(minbucket = 7,
+#' mincriterion = 0.95))
+#' tree <- list(tree = datact, given_ind = given_ind, dependent_ind = dependent_ind)
+#' shapr:::sample_ctree(tree = tree, n_samples = n_samples, x_test = x_test_dt, x_train = x_train,
+#' p = length(x_test), sample = TRUE)
+#'
+#' @author Annabelle Redelmeier
+sample_ctree <- function(tree,
+                         n_samples,
+                         x_test,
+                         x_train,
+                         p,
+                         sample) {
+  datact <- tree$tree
+  using_partykit <- (class(datact)[1] != "BinaryTree")
+
+  cnms <- colnames(x_test)
+  if (length(tree$given_ind) %in% c(0, p)) {
+    ret <- x_test
+  } else {
+    given_ind <- tree$given_ind
+
+    dependent_ind <- tree$dependent_ind
+
+    x_test_given <- x_test[,
+                           given_ind,
+                           drop = FALSE,
+                           with = FALSE] #
+    xp <- x_test_given
+    colnames(xp) <- paste0("V", given_ind) # this is important for where() below
+
+    if (using_partykit) {
+      fit.nodes <- predict(object = datact,
+                           type = "node")
+      # newdata must be data.frame + have the same colnames as x
+      pred.nodes <- predict(object = datact, newdata = xp,
+                            type = "node")
+
+    } else {
+      fit.nodes <- party::where(object = datact)
+      # newdata must be data.frame + have the same colnames as x
+      pred.nodes <- party::where(object = datact, newdata = xp)
+    }
+
+    rowno <- 1:dim(x_train)[1]
+
+    if (!sample) {
```

Comment:
I know we have talked about this before but this "sample" functionality seems unusually complicated for something that is not very useful (and probably only we care about it). Perhaps in another version we think about removing "sample"

Reply:
@aredelmeier Please take a careful look at the new version. I think that looks better and should be doing the same. All tests are passing.

Reply:
This looks good to me.